src/flatfile.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <flatfile.h>
+#include <logging.h>
+#include <tinyformat.h>
+#include <util/system.h>
+
+std::string FlatFilePos::ToString() const
+{
+    return strprintf("FlatFilePos(nFile=%i, nPos=%i)", nFile, nPos);
+}
+
+FlatFileSeq::FlatFileSeq(fs::path dir, const char* prefix, size_t chunk_size)
+    : m_dir(std::move(dir)), m_prefix(prefix), m_chunk_size(chunk_size)
+{}
+
+fs::path FlatFileSeq::FileName(const FlatFilePos& pos) const
+{
+    return m_dir / strprintf("%s%05u.dat", m_prefix, pos.nFile);
+}
+
+FILE* FlatFileSeq::Open(const FlatFilePos& pos, bool read_only)
+{
+    if (pos.IsNull())
+        return nullptr;
+    fs::path path = FileName(pos);
+    fs::create_directories(path.parent_path());
+    FILE* file = fsbridge::fopen(path, read_only ? "rb": "rb+");
+    if (!file && !read_only)
+        file = fsbridge::fopen(path, "wb+");
+    if (!file) {
+        LogPrintf("Unable to open file %s\n", path.string());
+        return nullptr;
+    }
+    if (pos.nPos) {
+        if (fseek(file, pos.nPos, SEEK_SET)) {
+            LogPrintf("Unable to seek to position %u of %s\n", pos.nPos, path.string());
+            fclose(file);
+            return nullptr;
+        }
+    }
+    return file;
+}
+
+size_t FlatFileSeq::Allocate(const FlatFilePos& pos, size_t add_size, bool& out_of_space)
+{
+    out_of_space = false;
+
+    unsigned int n_old_chunks = (pos.nPos + m_chunk_size - 1) / m_chunk_size;
+    unsigned int n_new_chunks = (pos.nPos + add_size + m_chunk_size - 1) / m_chunk_size;
+    if (n_new_chunks > n_old_chunks) {
+        size_t old_size = pos.nPos;
+        size_t new_size = n_new_chunks * m_chunk_size;
+        size_t inc_size = new_size - old_size;
+
+        if (CheckDiskSpace(m_dir, inc_size)) {
```

Comment:
Any concerns about the type mismatch here between `inc_size` (`size_t`) and the second argument to `CheckDiskSpace` which is a `uint64_t`?

Reply:
Do you get a compiler warning? If not, the implicit cast ought to be fine.

Reply:
No compiler warning, maybe just a note for the future that the sizes of the integer types should probably be reconciled at some point.